core/src/main/java/org/elasticsearch/common/lucene/search/MatchNoDocsQuery.java

```diff
@@ -0,0 +1,55 @@
+package org.elasticsearch.common.lucene.search;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.*;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Set;
+
+/**
+ * A query that matches no documents and prints the reason why in the toString method.
+ */
+public class MatchNoDocsQuery extends Query {
+    /**
+     * The reason why the query does not match any document.
+     */
+    private final String reason;
+
+    public MatchNoDocsQuery(String reason) {
+        this.reason = reason;
+    }
+
+    public Weight createWeight(IndexSearcher searcher, boolean needsScores) throws IOException {
+        return new Weight(this) {
+            @Override
+            public void extractTerms(Set<Term> terms) {
+            }
+
+            @Override
+            public Explanation explain(LeafReaderContext context, int doc) throws IOException {
+                return Explanation.noMatch(reason);
+            }
+
+            @Override
+            public float getValueForNormalization() throws IOException {
+                return 0;
```

Comment:
Since the purpose is to replace eg. range queries, I am wondering that we might want normalization to behave like range queries so that scoring is the same (that would come ootb if we extend ConstantScoreWeight and  don't extend getValueForNormalization/normalize). I realize this is not really related to this PR since we already have the issue today in master since we use Lucene's MatchNoDocsQuery which also returns a norm of zero, so maybe it would best be addressed in a follow-up PR? (do as you prefer)